mithril-signer/src/services/certifier.rs

```diff
@@ -0,0 +1,374 @@
+            })
+            .await
+            .unwrap();
+
+        let signed_beacons = signed_beacons_store.signed_beacons().await;
+        assert_eq!(
+            vec![SignedEntityType::MithrilStakeDistribution(Epoch(4))],
+            signed_beacons
+        );
+    }
+
+    #[tokio::test]
+    async fn if_already_signed_a_beacon_is_not_returned_anymore() {
+        let ticker_service = DumbTickerService::new(TimePoint::new(1, 14, ChainPoint::dummy()));
+        let certifier_service = SignerCertifierService::new(
+            Arc::new(ticker_service),
+            Arc::new(DumbSignedBeaconStore::default()),
+            Arc::new(DumbSignedEntityConfigProvider::new(
+                CardanoTransactionsSigningConfig::dummy(),
+                SignedEntityTypeDiscriminants::all(),
+            )),
+            Arc::new(SignedEntityTypeLock::new()),
+        );
+
+        let first_beacon_to_sign = certifier_service
+            .get_beacon_to_sign()
+            .await
+            .unwrap()
+            .unwrap();
+        certifier_service
+            .mark_beacon_as_signed(&first_beacon_to_sign.clone())
+            .await
+            .unwrap();
+        let second_beacon_to_sign = certifier_service
+            .get_beacon_to_sign()
+            .await
+            .unwrap()
+            .unwrap();
+
+        assert_ne!(
+            first_beacon_to_sign.signed_entity_type,
+            second_beacon_to_sign.signed_entity_type
+        );
+    }
+
+    #[tokio::test]
+    async fn draining_out_all_beacon_to_signs_dont_repeat_value_and_use_signed_entity_discriminant_order(
+    ) {
+        let ticker_service = DumbTickerService::new(TimePoint::new(1, 14, ChainPoint::dummy()));
+        let certifier_service = SignerCertifierService::new(
+            Arc::new(ticker_service),
+            Arc::new(DumbSignedBeaconStore::default()),
+            Arc::new(DumbSignedEntityConfigProvider::new(
+                CardanoTransactionsSigningConfig::dummy(),
+                SignedEntityTypeDiscriminants::all(),
+            )),
+            Arc::new(SignedEntityTypeLock::new()),
+        );
+
+        let mut previous_beacon_to_sign = certifier_service
+            .get_beacon_to_sign()
+            .await
+            .unwrap()
+            .expect("There should be a beacon to sign since nothing is locked or signed");
+        let mut all_signed_beacons = vec![previous_beacon_to_sign.clone()];
+
+        loop {
+            certifier_service
+                .mark_beacon_as_signed(&previous_beacon_to_sign)
+                .await
+                .unwrap();
+            let next_beacon_to_sign = certifier_service.get_beacon_to_sign().await.unwrap();
+
+            if let Some(beacon) = next_beacon_to_sign {
+                assert!(
+                    !all_signed_beacons.contains(&beacon),
+                    "Beacon should not repeat"
+                );
+                assert!(
+                    SignedEntityTypeDiscriminants::from(
```

Comment:
We can try to split this test in two to see if it will be simplest to understand
One to check only the order and another one for the rest.